### DISCUSSION TITLE
Can we use mill to build a stand-alone launcher

### DISCUSSION THREAD

**Discussion by hmf** [almond-sh/almond#923] [Q&A] [answered]
Is there any way we can use the mill command to build an independent (fat) Jar for a launcher just as we do with coursier?

TIA

**Comment by alexarchambault** [comment, marked as answer]
The `launcher` task (described a bit [here](https://almond.sh/docs/dev-from-sources) in particular) should generate something like that: a launcher that can be copied to other machines, even though it doesn't embed all of its dependencies (it downloads them via the coursier cache upon first launch).

**Comment by alexarchambault** [comment]
It might be possible to embed all dependencies in the generated launcher by tweaking `createLauncher` in `project/settings.sc`, but I don't recall the exact details of it right now…

**Reply by hmf** [reply]
Did not see references to Mill's assembly. I do see a `.withHybridAssembly(true)` though. When I get the chance, I will try again. Thank you.

**Reply by alexarchambault** [reply]
Assemblies don't work fine for us here. The `createLauncher` generates a coursier "bootstrap", that loads the Almond class path in two class loaders (dependencies shared with users, that is the class path of scala-kernel-api, in a first loader, and the remaining JARs in a second loader, with the first one as parent). Later on Almond passes the first class loader to Ammonite, and users only see the JARs loaded by it, not the Almond internal dependencies loaded in the second class loader. That lowers the odds of conflicts between Almond's internal dependencies (which include specific versions of fs2, cats, and cats-effect) and the dependencies loaded by users.

Assemblies somehow work at first, but don't isolate dependencies this way.

**Reply by hmf** [reply]
Thanks for the explanation.
